### PR TITLE
EZP-24744 - Increase password security

### PR DESCRIPTION
**Pull request by glye**
> https://jira.ez.no/browse/EZP-24744
> Based on https://github.com/ezsystems/ezpublish-kernel/pull/1592
> Same for legacy: https://github.com/ezsystems/ezpublish-legacy/pull/1322
> Ready for review

Adapted to current master
Added `PASSWORD_HASH_PHP_DEFAULT` in addition to `PASSWORD_HASH_BCRYPT`
Set `PASSWORD_HASH_PHP_DEFAULT` as default
Moved doc to 6.12 file

### CODE REVIEW

**Comment by andrerom** [comment]
Nice! You can btw change the PR to master, that's 6.12 (and this is a new feature so it's for 6.12).

**Comment by andrerom** [comment]
Besides that, will you give the legacy PR same threatment? Will need to that in order to merge these two (for legacy bridge use), if so the conflict in legacy should be increased to `<6.12 || >=2014.11 <2017.10`: https://github.com/ezsystems/ezpublish-legacy/blob/master/composer.json#L67

**Comment by andrerom** [comment]
Also, while at it maybe we should add constant for argon2: https://wiki.php.net/rfc/argon2_password_hash

**Comment by glye** [comment]
Yes, working on legacy, will see about argon2.

**Comment by andrerom** [comment]
> will see about argon2.

as last part now reflects, I don't think it is the right time to add it. It will complicate doc on this, better to add alter when more widely available (currently not at all).

**Comment by glye** [comment]
AppVeyor's problem is it cannot download php-7.1.2-nts-Win32-VC14-x86.zip
This didn't fail on 6.11, but I guess it uses another PHP version.

**Comment by alongosz** [comment]
> AppVeyor's problem is it cannot download php-7.1.2-nts-Win32-VC14-x86.zip. This didn't fail on 6.11, but I guess it uses another PHP version.

I'll take a look tomorrow. Handling of Windows binaries for PHP is a mess caused by their maintainer ;)

**Comment by alongosz** [comment]
@glye This time AppVeyor failure looks like some random unavailability of server with downloads. Unfortunately I can't restart build, has to be triggered by force push I think.

**Comment by glye** [comment]
@alongosz Thanks. I have pushed, let's hope the force is with me.

**Comment by andrerom** [comment]
@glye I guess we should update dat dump with this as well? And I guess we need to ping web team for also doing that on demo installs _(not a must, not doing it can pherhaps also be a good thing there to showcase hash being updated to the new format on use)_.

**Comment by glye** [comment]
Updated data dump, and a test. Unsure how many tests need to be changed because of this, let's see.

File: eZ/Publish/API/Repository/Tests/FieldType/UserIntegrationTest.php [hunk 1]
```diff
@@ -132,8 +132,7 @@ public function assertFieldDataLoadedCorrect(Field $field)
             'hasStoredLogin' => true,
             'login' => 'hans',
             'email' => 'hans@example.com',
-            'passwordHash' => '680869a9873105e365d39a6d14e68e46',
-            'passwordHashType' => 2,
+            'passwordHashType' => 7,
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/FieldType/UserIntegrationTest.php]: Nitpick: I would prefer here a reference to constant.

File: eZ/Publish/Core/FieldType/Tests/Integration/User/UserStorage/UserStorageGatewayTest.php [hunk 2]
```diff
@@ -31,8 +31,8 @@ public function providerForGetFieldData()
                 'contentId' => 10,
                 'login' => 'anonymous',
                 'email' => 'nospam@ez.no',
-                'passwordHash' => '4e6f6184135228ccd45f8233d72a0363',
-                'passwordHashType' => '2',
+                'passwordHash' => '$2y$10$35gOSQs6JK4u4whyERaeUuVeQBi2TUBIZIfP7HEj7sfz.MxvTuOeC',
+                'passwordHashType' => '7',
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Tests/Integration/User/UserStorage/UserStorageGatewayTest.php]: Same nitpick about using constant :)

File: eZ/Publish/Core/FieldType/Tests/Integration/User/UserStorage/UserStorageGatewayTest.php [hunk 3]
```diff
@@ -41,8 +41,8 @@ public function providerForGetFieldData()
                 'contentId' => 14,
                 'login' => 'admin',
                 'email' => 'spam@ez.no',
-                'passwordHash' => 'c78e3b0f3d9244ed8c6d1c29464bdff9',
-                'passwordHashType' => '2',
+                'passwordHash' => '$2y$10$FDn9NPwzhq85cLLxfD5Wu.L3SL3Z/LNCvhkltJUV0wcJj7ciJg2oy',
+                'passwordHashType' => '7',
```
Comment (alongosz) [inline eZ/Publish/Core/FieldType/Tests/Integration/User/UserStorage/UserStorageGatewayTest.php]: and the same here ;)

**Comment by alongosz** [comment]
@damianz5: you probably need to align demo data in the future (looks like now should be BC) :)

Note to self: update #1941

**Comment by glye** [comment]
Thanks @alongosz, updated.

`UserStorageGatewayTest` may fail now if `assertEquals()` does strict comparison, in which case we can cast the data to string.

**Pull request merged**